tutorials/backend/hasura-authentication/tutorial-site/content/integrations/casdoor.md

```diff
@@ -0,0 +1,149 @@
+
+Save the user after adding a username and the organization Hasura (other details are optional).
+
+Now you need to set up a password for your user, which you can do by clicking "Manage Your Password."
+
+Choose a password for your user and confirm it.
+
+## Create a Hasura instance
+
+Create a Hasura instance using either
+[Hasura Cloud](https://hasura.io/docs/latest/getting-started/getting-started-cloud/) or
+[Docker](https://hasura.io/docs/latest/getting-started/docker-simple/).
+
+Now, create a `users` table with the following columns:
+
+- `id` of type Text (Primary Key)
+- `username` of type Text
+
+See the image below for reference.
+
+![Picture showing how to create a table in Hasura](https://graphql-engine-cdn.hasura.io/learn-hasura/assets/graphql-hasura-authentication/keycloak/hasura-create-table.png)
+
+The next step is to create a `user` role for the app. Users should be able to see only their records, but not the other
+people’s records.
+
+Configure the `user` role as shown in the image below. For more information, read about
+[configuring permission rules in Hasura](https://hasura.io/docs/latest/graphql/core/auth/authorization/permission-rules/).
+
+![Picture showing how to set permissions in Hasura](https://graphql-engine-cdn.hasura.io/learn-hasura/assets/graphql-hasura-authentication/keycloak/hasura-set-permissions.png)
+
+This way, users can only access their own records.
+
+For testing purposes, add a dummy user. This is to ensure that when you use the JWT token, you only see your user’s
+details and not other users’ details.
+
+![Picture showing how to add a table record in Hasura](https://graphql-engine-cdn.hasura.io/learn-hasura/assets/graphql-hasura-authentication/keycloak/hasura-dummy-user.png)
+
+Next, set the `JWT_SECRET` in Hasura.
+
+## Configure Hasura with Casdoor
+
+In this step, you need to add the **HASURA_GRAPHQL_JWT_SECRET** to Hasura.
+
+To do so, go to the Hasura docker-compose.yaml and then add the new `HASURA_GRAPHQL_JWT_SECRET` as below.
+
+The `HASURA_GRAPHQL_JWT_SECRET` should be in the following format:
+
+```
+HASURA_GRAPHQL_JWT_SECRET: '{"claims_map": {
+      "x-hasura-allowed-roles": ["user","editor"],
+      "x-hasura-default-role": "user",
+      "x-hasura-user-id": "userID"
+    },"jwk_url":"https://door.casdoor.com/.well-known/jwks"}'
+```
+
+Save the change, and restart the Docker container.
+
+![Add Casdoor JWT URL to Hasura](https://github.com/RanTao123/image/blob/main/MD$GWN%5BBET2O538TG~LNZIM.png?raw=true)
+
+## Retrieve JWT Token
+
+Since there is no client implementation, you can get your access token by making a request by below URL:
+
+```
+http://localhost:8000/login/oauth/authorize?client_id=<client ID>>&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A8080%2Flogin&scope=read&state=app-built-in<public certificate>>
+```
+
+Change the client ID to the ID you copied before, and input the public certificate of Casdoor, which you can find in
+`Casdoor/Certs`.
```

Comment:
For me, this directory wasn't available in the repository. Instead, a user needs to navigate to [localhost:7001/certs](http://localhost:7001/certs) to retrieve the public certificate.